{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copyright **`(c)`** 2023 Giovanni Squillero `<giovanni.squillero@polito.it>`  \n",
    "[`https://github.com/squillero/computational-intelligence`](https://github.com/squillero/computational-intelligence)  \n",
    "Free for personal or classroom use; see [`LICENSE.md`](https://github.com/squillero/computational-intelligence/blob/master/LICENSE.md) for details.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LAB9\n",
    "\n",
    "Wrote a local-search algorithm (eg. an EA) able to solve the *Problem* instances 1, 2, 5, and 10 on a 1000-loci genomes, using a minimum number of fitness calls. That's all.\n",
    "\n",
    "### Deadlines:\n",
    "\n",
    "* Submission: Sunday, December 3 ([CET](https://www.timeanddate.com/time/zones/cet))\n",
    "* Reviews: Sunday, December 10 ([CET](https://www.timeanddate.com/time/zones/cet))\n",
    "\n",
    "Notes:\n",
    "\n",
    "* Reviews will be assigned  on Monday, December 4\n",
    "* You need to commit in order to be selected as a reviewer (ie. better to commit an empty work than not to commit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import choices\n",
    "from dataclasses import dataclass\n",
    "import lab9_lib\n",
    "from random import random, choice, randint\n",
    "from copy import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "TOURNAMENT_SIZE = 2\n",
    "MUTATION_PROBABILITY = .15\n",
    "POPULATION_SIZE= 100\n",
    "PROBLEM_DIMENSION = 2\n",
    "GENO_LENGHT = 1000\n",
    "OFFSPRING_SIZE=20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "@dataclass\n",
    "class Individual:\n",
    "    genotype: list[int]\n",
    "    fitness: float\n",
    "\n",
    "population = [\n",
    "    Individual(\n",
    "        genotype=choices([0, 1], k = GENO_LENGHT),\n",
    "        fitness=None,\n",
    "    )\n",
    "    for _ in range(POPULATION_SIZE)\n",
    "]\n",
    "\n",
    "fitness = lab9_lib.make_problem(PROBLEM_DIMENSION)\n",
    "for i in population:\n",
    "    i.fitness = fitness(i.genotype)\n",
    "\n",
    "def select_parent(pop):\n",
    "    pool=[choice(pop) for _ in range(TOURNAMENT_SIZE)] #list of TOURNAMENT_SIZE individuals and I have to pick the best\n",
    "    champion = max(pool, key=lambda i: i.fitness) #best individual in the pool. Labmda is a callable object on which I gets individual to call on it the fitness function\n",
    "    return champion\n",
    "\n",
    "def mutate(ind: Individual) -> Individual:\n",
    "    offspring=copy(ind)\n",
    "    pos = randint(0, GENO_LENGHT-1) #I mutate a random gene\n",
    "    offspring.genotype[pos] = not offspring.genotype[pos]\n",
    "    return offspring\n",
    "\n",
    "\n",
    "#crossover ricombination    \n",
    "def one_cut_cover(ind1: Individual,ind2: Individual) ->Individual:\n",
    "    cut_point=randint(0,GENO_LENGHT-1)\n",
    "    offspring=Individual(fitness=None,genotype=ind1.genotype[:cut_point]+ind2.genotype[cut_point:])\n",
    "    assert len(offspring.genotype) == GENO_LENGHT\n",
    "    return offspring\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fitness = lab2_lib.make_problem(10)\n",
    "# for n in range(10):\n",
    "#     ind = choices([0, 1], k = 50)\n",
    "#     print(''.join(str(g) for g in ind), ':', fitness(ind))\n",
    "\n",
    "# print(fitness.calls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5106\n",
      "0.5118\n",
      "0.5118\n",
      "0.5118\n",
      "0.5152\n",
      "0.517\n",
      "0.517\n",
      "0.5296000000000001\n",
      "0.5296000000000001\n",
      "0.5316000000000001\n",
      "0.5316000000000001\n",
      "0.5342\n",
      "0.5438\n",
      "0.551\n",
      "0.551\n",
      "0.551\n",
      "0.551\n",
      "0.5536\n",
      "0.5614\n",
      "0.5614\n",
      "0.5664\n",
      "0.5664\n",
      "0.5664\n",
      "0.5714\n",
      "0.5714\n",
      "0.5714\n",
      "0.5714\n",
      "0.5714\n",
      "0.5746\n",
      "0.5746\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5746\n",
      "0.5774\n",
      "0.5774\n",
      "0.583\n",
      "0.583\n",
      "0.583\n",
      "0.583\n",
      "0.583\n",
      "0.583\n",
      "0.583\n",
      "0.583\n",
      "0.583\n",
      "0.583\n",
      "0.5868\n",
      "0.5868\n",
      "0.5868\n",
      "0.5868\n",
      "0.5868\n",
      "0.5868\n",
      "0.5933999999999999\n",
      "0.5936\n",
      "0.5936\n",
      "0.5936\n",
      "0.595\n",
      "0.595\n",
      "0.5956\n",
      "0.5966\n",
      "0.5966\n",
      "0.5966\n",
      "0.5966\n",
      "0.5972000000000001\n",
      "0.5972000000000001\n",
      "0.5972000000000001\n",
      "0.5972000000000001\n",
      "0.5972000000000001\n",
      "0.5978\n",
      "0.5983999999999999\n",
      "0.5983999999999999\n",
      "0.5983999999999999\n",
      "0.5998\n",
      "0.5998\n",
      "0.5998\n",
      "0.5998\n",
      "0.6006\n",
      "0.6008\n",
      "0.601\n",
      "0.603\n",
      "0.603\n",
      "0.603\n",
      "0.6043999999999999\n",
      "0.6043999999999999\n",
      "0.6056\n",
      "0.6058\n",
      "0.6058\n",
      "0.6076\n",
      "0.6076\n",
      "0.6076\n",
      "0.6076\n",
      "0.6076\n",
      "0.6076\n",
      "0.6078\n",
      "0.6098\n",
      "0.6098\n",
      "0.6116\n",
      "0.6116\n",
      "0.6136\n",
      "0.6136\n",
      "0.6136\n",
      "0.6137999999999999\n",
      "0.6137999999999999\n",
      "2100\n"
     ]
    }
   ],
   "source": [
    "for generation in range(100): #we have to take parents from population\n",
    "    offspring = list()\n",
    "    for counter in range(OFFSPRING_SIZE):\n",
    "        if random() < MUTATION_PROBABILITY:  #self adapt mutation probability\n",
    "            #do_ random   #add more clever mutation\n",
    "            p=select_parent(population) #in case of mutation I take one parent\n",
    "            o=mutate(p)\n",
    "        else:\n",
    "            #crossover #add more xovers\n",
    "            p1=select_parent(population) #in case of crossover I take two parents\n",
    "            p2=select_parent(population) \n",
    "            o= one_cut_cover(p1,p2)\n",
    "        offspring.append(o)\n",
    "        #Now we have to evaluate offspring\n",
    "    for i in offspring:\n",
    "        i.fitness = fitness(i.genotype)\n",
    "    population.extend(offspring) # I merged new individuals in population\n",
    "    population.sort(key=lambda i:i.fitness,reverse=True) \n",
    "    population = population[:POPULATION_SIZE] #I'm cutting part of the population , i'm taking only the best individuals, this is named survival selection   \n",
    "    print(population[0].fitness)\n",
    "print(fitness.calls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ci-fLJ3OwGs-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
